File: game_app/src/main/cpp/GameEngine.cpp
class TicTacToeEngine {
private:
    char board[3][3];
    int moveCount;

public:
    TicTacToeEngine() {
        initBoard();
    }

    void initBoard() {
        for (int i=0; i<3; ++i) {
            for (int j=0; j<3; ++j) {
                board[i][j] = '-';
            }
        }
        moveCount = 0;
    }


    bool placeMarker(int row, int col, char player) {
        if (row < 0 || row > 2 || col < 0 || col > 2 ) {
            return false;
        }
        if (board[row][col] != '-') {
            return false;
        }
        board[row][col] = player;
        moveCount++;
        return true;
    }

    char getCell(int row, int col) {
        if (row < 0 || row > 2 || col < 0 || col > 2 ) {
            return '-';
        }
        return board[row][col];
    }

    char checkWinner() {
        for (int i=0; i<3; ++i) {
            if (board[i][0] != '-' &&
                board[i][1] == board[i][0] &&
                board[i][2] == board[i][1]) 
            {
                return board[i][0];
            }
        }
        for (int i=0; i<3; ++i) {
            if (board[0][i] != '-' &&
                board[1][i] == board[0][i] &&
                board[2][i] == board[1][i]) 
            {
                return board[0][i];
            }
        }

        if (board[0][0] != '-' &&
            board[1][1] == board[0][0] &&
            board[2][2] == board[1][1])
        {
            return board[0][0];
        }

        if (board[2][0] != '-' &&
            board[1][1] == board[2][0] &&
            board[0][2] == board[1][1])
        {
            return board[2][0];
        }

        return '-';
    }

    int getMoveCount() {
        return moveCount;
    }

    void reset() {
        initBoard();
    }

    bool isGameOver()  {
        if (moveCount == 9) return true;
        if (checkWinner() != '-') return true;
        return false;
    }

};

static TicTacToeEngine* g_engine = nullptr;


void initEngine() {
    if (g_engine == nullptr) {
        g_engine = new TicTacToeEngine();
    } else {
        g_engine->reset();
    }
}

TicTacToeEngine* getEngine() {
    if (g_engine == nullptr) {
        initEngine();
    }
    return g_engine;
}


bool placeMarker(int row, int col, char player) {
    TicTacToeEngine* engine = getEngine();
    bool result = g_engine->placeMarker(row, col, player);
    return result;
}

void resetGame() {
    g_engine->reset();
}

char checkWinner() {
    return g_engine->checkWinner();
}

char getCell(int row, int col) {
    return g_engine->getCell(row, col);
}

bool isGameOver() {
    return g_engine->isGameOver();
}

int getMoveCount() {
    return g_engine->getMoveCount();
}

File: game_app/src/main/java/game/GameEngine.java
package game;

import java.io.File;

public class GameEngine {
    
    static {
        try {
            // Próba załadowania z target/classes/native (dla mvn exec:java)
            String libraryPath = "target/classes/native/libtictactoe_engine.so";
            File libFile = new File(libraryPath);
            
            if (!libFile.exists()) {
                // Próba załadowania z resources (dla JAR)
                libraryPath = "src/main/resources/native/libtictactoe_engine.so";
                libFile = new File(libraryPath);
            }
            
            if (libFile.exists()) {
                System.load(libFile.getAbsolutePath());
                System.out.println("Biblioteka natywna załadowana z: " + libFile.getAbsolutePath());
            } else {
                throw new UnsatisfiedLinkError("Nie znaleziono biblioteki libtictactoe_engine.so");
            }
        } catch (Exception e) {
            e.printStackTrace();
            throw new UnsatisfiedLinkError("Błąd ładowania biblioteki: " + e.getMessage());
        }
    }
    
    public native void initGame();
    public native boolean placeMarker(int row, int col, char player);
    public native char checkWinner();
    public native char getCell(int row, int col);
    public native void resetGame();
    public native boolean isGameOver();
    public native int countMoves();
}
